src/DataStreams/RemoteQueryExecutorReadContext.h

```diff
@@ -0,0 +1,263 @@
+            if (events[i].data.fd == timer.getDescriptor())
+                has_timer_alarm = true;
+            if (events[i].data.fd == pipe_fd[0])
+                is_pipe_alarmed = true;
+        }
+
+        if (is_pipe_alarmed)
+            return false;
+
+        if (has_timer_alarm && !is_socket_ready)
+        {
+            /// Socket receive timeout. Drain it in case or error, or it may be hide by timeout exception.
+            timer.drain();
+            throw NetException("Timeout exceeded", ErrorCodes::SOCKET_TIMEOUT);
+        }
+
+        return true;
+    }
+
+    void setTimer() const
+    {
+        /// Did not get packet yet. Init timeout for the next async reading.
+        timer.reset();
+
+        if (receive_timeout.totalMicroseconds())
+            timer.setRelative(receive_timeout);
+    }
+
+    bool resumeRoutine()
+    {
+        if (is_read_in_progress && !checkTimeout())
+            return false;
+
+        {
+            std::lock_guard guard(fiber_lock);
+            if (!fiber)
+                return false;
+
+//            if (!connection_lock->owns_lock())
+//                connection_lock->lock();
+
+            fiber = std::move(fiber).resume();
+
+//            if (!is_read_in_progress)
+//                connection_lock->unlock();
+        }
+
+        if (exception)
+            std::rethrow_exception(std::move(exception));
+
+        if (is_read_in_progress)
+        {
+            auto & socket = connections.getSocket();
+            try
+            {
+                setSocket(socket);
+            }
+            catch (DB::Exception & e)
+            {
+                e.addMessage(" while reading from socket ({})", socket.peerAddress().toString());
+                throw;
+            }
+        }
+
+        return true;
+    }
+
+    void cancel()
+    {
+        std::lock_guard guard(fiber_lock);
+        boost::context::fiber to_destroy = std::move(fiber);
+
+        uint64_t buf = 0;
+        while (-1 == write(pipe_fd[1], &buf, sizeof(buf)))
+        {
+            if (errno == EAGAIN)
+                break;
+
+            if (errno != EINTR)
+                throwFromErrno("Cannot write to pipe", ErrorCodes::CANNOT_READ_FROM_SOCKET);
+        }
+    }
+
+    ~RemoteQueryExecutorReadContext()
+    {
+        /// socket_fd is closed by Poco::Net::Socket
+        /// timer_fd is closed by TimerDescriptor
+        close(epoll_fd);
+    }
+
+    struct Routine
+    {
+        MultiplexedConnections & connections;
+        Self & read_context;
+        std::unique_ptr<std::unique_lock<std::mutex>> connection_lock;
+
+        Fiber operator()(Fiber && sink) const
+        {
+            try
+            {
+                while (true)
+                {
+                    read_context.is_read_in_progress = true;
+                    read_context.packet = connections.receivePacketUnlocked(&sink);
+                    read_context.is_read_in_progress = false;
```

Comment:
Never thought I'd suggest adding a callback instead of removing it, but there's always the first time: did you consider injecting a callback into the connection::read, that would set this flag, and also save the corresponding descriptor? You're injecting the fiber there anyway.

This control flow I can barely understand... At least with callback you'll be saving all the resume data (socket, flag, fiber) in one place, not in three different ones.

Reply:
Hm, ok. I will try it.